### PR TITLE
Add Bilinear Tensor Product operator.

### PR DESCRIPTION
**Pull request by peterzhang2029**
resolve https://github.com/PaddlePaddle/Paddle/issues/4789

### CODE REVIEW

File: paddle/operators/bilinear_tensor_product_op.cc [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    PADDLE_ENFORCE(ctx->HasInput("Y"), "Input(Y) should not be null.");
+    PADDLE_ENFORCE(ctx->HasInput("Weight"),
+                   "Input(Weight) should not be null.");
+    PADDLE_ENFORCE(ctx->HasOutput("Out"), "Output(Out) should not be null.");
+    auto x_dims = ctx->GetInputDim("X");
+    auto y_dims = ctx->GetInputDim("Y");
+    auto weight_dims = ctx->GetInputDim("Weight");
+
+    PADDLE_ENFORCE_EQ(x_dims.size(), 1, "The input X must be a vector.");
+    PADDLE_ENFORCE_EQ(y_dims.size(), 1, "The input Y must be a vector.");
```
Comment (lcy-seso) [inline paddle/operators/bilinear_tensor_product_op.cc]: Why X and Y are vectors? A mini-batch is a 2-D tensor. The 1st dimension is batch size, and the 2nd dimension is the hidden size of X and Y.

Comment (peterzhang2029) [reply paddle/operators/bilinear_tensor_product_op.cc]: Done

File: paddle/operators/bilinear_tensor_product_op.cu [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+#define EIGEN_USE_GPU
+#include "paddle/operators/bilinear_tensor_product_op.h"
+
+namespace paddle {
+namespace operators {
+
+template <typename Place, typename T>
+class BilinearTensorProductCUDAKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& ctx) const override {
```
Comment (lcy-seso) [inline paddle/operators/bilinear_tensor_product_op.cu]: 这里GPU下为什么需要从CPU拷贝输入输出呢？不管是Eigen还是矩阵乘法都有GPU下的实现。

File: paddle/operators/bilinear_tensor_product_op.cc [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+  }
+};
+
+class BilinearTensorProductOpGrad : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+ protected:
+  void InferShape(framework::InferShapeContext* ctx) const override {
+    PADDLE_ENFORCE(ctx->HasInput("X"), "Input(X) should not be null");
```
Comment (lcy-seso) [inline paddle/operators/bilinear_tensor_product_op.cc]: Input(X) should not be null 注释的后面加上句号~ 下同。

File: paddle/operators/bilinear_tensor_product_op.cc [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+                   "Input (Out@GRAD) should not be null");
+    auto x_dims = ctx->GetInputDim("X");
+    auto y_dims = ctx->GetInputDim("Y");
+    auto weight_dims = ctx->GetInputDim("Weight");
+    auto out_dims = ctx->GetInputDim(framework::GradVarName("Out"));
+
+    PADDLE_ENFORCE_EQ(out_dims.size(), 2, "The Out@GRAD must be a 2D Tensor.");
+    PADDLE_ENFORCE_EQ(
+        x_dims[0], out_dims[0],
+        "The first dimension(batch_size) of Out@GRAD must be equal with "
```
Comment (lcy-seso) [inline paddle/operators/bilinear_tensor_product_op.cc]: be equal to

File: paddle/operators/bilinear_tensor_product_op.cc [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+        "the first dimension of the X.");
+    PADDLE_ENFORCE_EQ(weight_dims[0], out_dims[1],
+                      "The second dimension of Out@GRAD must be equal with "
```
Comment (lcy-seso) [inline paddle/operators/bilinear_tensor_product_op.cc]: be equal to

File: paddle/operators/bilinear_tensor_product_op.cc [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+                      "the third dimension of the Weight.");
```
Comment (lcy-seso) [inline paddle/operators/bilinear_tensor_product_op.cc]: Weight --> Input(Weight)

File: paddle/operators/bilinear_tensor_product_op.cc [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+
+    if (ctx->HasInput("Bias")) {
+      auto bias_dims = ctx->GetInputDim("Bias");
+      PADDLE_ENFORCE_EQ(bias_dims[1], out_dims[1],
+                        "The second dimension of Bias must be equal with "
```
Comment (lcy-seso) [inline paddle/operators/bilinear_tensor_product_op.cc]: be equal to

File: paddle/operators/bilinear_tensor_product_op.h [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+#pragma once
+
+#include "paddle/framework/eigen.h"
+#include "paddle/framework/op_registry.h"
+#include "paddle/operators/math/math_function.h"
+
+namespace paddle {
+namespace operators {
+
+using Tensor = framework::Tensor;
```
Comment (lcy-seso) [inline paddle/operators/bilinear_tensor_product_op.h]:
```cpp
using framework::LoDTensor;
```

File: paddle/operators/bilinear_tensor_product_op.h [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+    auto weight_dims = weight->dims();
+    auto place = ctx.GetEigenDevice<Place>();
+
+    // Create the temporary variables.
+    Tensor left_mul;
+    left_mul.mutable_data<T>(framework::make_ddim({batch_size, weight_dims[2]}),
+                             ctx.GetPlace());
+    auto left_mul_mat = EigenMatrix<T>::From(left_mul);
+    Tensor output_col;
+    output_col.mutable_data<T>(framework::make_ddim({weight_dims[0]}),
```
Comment (lcy-seso) [inline paddle/operators/bilinear_tensor_product_op.h]: 一定需要这个临时变量吗？为什么不可以从输出Tensor里面切出一片，构造为一个 EigenMatrix，将69行的计算结果直接赋值给从输出Tensor切出来的一片。

File: paddle/operators/bilinear_tensor_product_op.cc [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+    auto y_dims = ctx->GetInputDim("Y");
+    auto weight_dims = ctx->GetInputDim("Weight");
+
+    PADDLE_ENFORCE_EQ(x_dims.size(), 2, "The input X must be a 2D Tensor.");
+    PADDLE_ENFORCE_EQ(y_dims.size(), 2, "The input Y must be a 2D Tensor.");
+    PADDLE_ENFORCE_EQ(weight_dims.size(), 3,
+                      "The input Weight must be a 3D tensor.");
+    PADDLE_ENFORCE_GT(weight_dims[0], 0,
+                      "The first dimension of Weight must be larger than 0.");
+    PADDLE_ENFORCE_GT(weight_dims[1], 0,
```
[the remark this answers is not in the document]
Comment (peterzhang2029) [reply paddle/operators/bilinear_tensor_product_op.cc]: Done

File: paddle/operators/bilinear_tensor_product_op.cc (hunk 3 above)
Comment (peterzhang2029) [reply paddle/operators/bilinear_tensor_product_op.cc]: Done

File: paddle/operators/bilinear_tensor_product_op.h (hunk 8 above)
Comment (peterzhang2029) [reply paddle/operators/bilinear_tensor_product_op.h]: 因为这个OP没有用到LoDTensor，所以改成了using framework::Tensor;

File: paddle/operators/bilinear_tensor_product_op.cc [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+    PADDLE_ENFORCE(ctx->HasInput("X"), "Input(X) should not be null.");
+    PADDLE_ENFORCE(ctx->HasInput("Y"), "Input(Y) should not be null.");
+    PADDLE_ENFORCE(ctx->HasInput("Weight"),
+                   "Input(Weight) should not be null.");
+    PADDLE_ENFORCE(ctx->HasOutput("Out"), "Output(Out) should not be null.");
+    auto x_dims = ctx->GetInputDim("X");
+    auto y_dims = ctx->GetInputDim("Y");
+    auto weight_dims = ctx->GetInputDim("Weight");
+
+    PADDLE_ENFORCE_EQ(x_dims.size(), 2, "The input X must be a 2D Tensor.");
```
[the remark this answers is not in the document]
Comment (peterzhang2029) [reply paddle/operators/bilinear_tensor_product_op.cc]: Done

File: paddle/operators/bilinear_tensor_product_op.cc [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+    PADDLE_ENFORCE_GT(weight_dims[2], 0,
+                      "The third dimension of Weight must be larger than 0.");
+    PADDLE_ENFORCE_EQ(x_dims[0], y_dims[0],
+                      "The first dimension(batch_size) of X must be "
+                      "equal with the first dimension of the Y.");
+    PADDLE_ENFORCE_EQ(x_dims[1], weight_dims[1],
+                      "The second dimension of X must be equal with the second "
+                      "dimension of the Weight.");
+    PADDLE_ENFORCE_EQ(y_dims[1], weight_dims[2],
+                      "The second dimension of Y must be equal with the third "
```
[the remark this answers is not in the document]
Comment (peterzhang2029) [reply paddle/operators/bilinear_tensor_product_op.cc]: Done

File: paddle/operators/bilinear_tensor_product_op.cc (hunk 4 above)
Comment (peterzhang2029) [reply paddle/operators/bilinear_tensor_product_op.cc]: Done

File: paddle/operators/bilinear_tensor_product_op.cc (hunk 5 above)
Comment (peterzhang2029) [reply paddle/operators/bilinear_tensor_product_op.cc]: Done

File: paddle/operators/bilinear_tensor_product_op.cc (hunk 6 above)
Comment (peterzhang2029) [reply paddle/operators/bilinear_tensor_product_op.cc]: Done

File: paddle/operators/bilinear_tensor_product_op.cc (hunk 7 above)
Comment (peterzhang2029) [reply paddle/operators/bilinear_tensor_product_op.cc]: Done

File: paddle/operators/bilinear_tensor_product_op.cu (hunk 2 above)
Comment (peterzhang2029) [reply paddle/operators/bilinear_tensor_product_op.cu]: 已经做了修正，不需要从CPU进行拷贝。

File: paddle/operators/bilinear_tensor_product_op.h (hunk 9 above)
Comment (peterzhang2029) [reply paddle/operators/bilinear_tensor_product_op.h]: Done

File: paddle/operators/bilinear_tensor_product_op.cc [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+    PADDLE_ENFORCE_EQ(y_dims[1], weight_dims[2],
+                      "The second dimension of Y must be equal to the third "
+                      "dimension of the Weight.");
+
+    if (ctx->HasInput("Bias")) {
+      auto bias_dims = ctx->GetInputDim("Bias");
+      PADDLE_ENFORCE_EQ(bias_dims.size(), 2UL,
+                        "The input Bias must have 2 dimensions.");
+      PADDLE_ENFORCE_EQ(bias_dims[0], 1UL,
+                        "The first dimention of input Bias must be 1.");
```
Comment (lcy-seso) [inline paddle/operators/bilinear_tensor_product_op.cc]: merge line 59 ~ 61

```cpp
 PADDLE_ENFORCE(bias_dims.size() == 2UL &&  bias_dims[1] == 1UL, 
               "The Input(bias) should be a 2-D tensor with the 2nd "
               "dimensions fixed to 1 (a row vector).")
```

File: paddle/operators/bilinear_tensor_product_op.cc [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+    ctx->ShareLoD("X", /*->*/ "Out");
+  }
+};
+
+class BilinearTensorProductOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  BilinearTensorProductOpMaker(framework::OpProto* proto,
+                               framework::OpAttrChecker* op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput("X", "The first input of BilinearTensorProduct op.");
```
Comment (lcy-seso) [inline paddle/operators/bilinear_tensor_product_op.cc]: I prefer to user bilinear_tensor_product operator. Because "BilinearTensorProductOp" is a name for the developer (in C++ codes), while "bilinear_tensor_product operator" is the name for the user (exposed by the user interface).

File: paddle/operators/bilinear_tensor_product_op.cc [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+    AddInput("Y", "The second input of BilinearTensorProduct op.");
+    AddInput("Weight", "The input weight of BilinearTensorProduct op.");
```
Comment (lcy-seso) [inline paddle/operators/bilinear_tensor_product_op.cc]: The learnable parameters of ...

File: paddle/operators/bilinear_tensor_product_op.cc [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+    AddInput("Bias", "The input bias of BilinearTensorProduct op.")
```
Comment (lcy-seso) [inline paddle/operators/bilinear_tensor_product_op.cc]: The learnable bias for bilinear_tensor_product operator. Do not use an abbreviation in the comments, if it is necessary (widely accept, or the name is too long).

File: paddle/operators/bilinear_tensor_product_op.cc [hunk 17]
```diff
@@ -0,0 +1,161 @@
...
+}  // namespace operators
+}  // namespace paddle
+
+namespace ops = paddle::operators;
+REGISTER_OP(bilinear_tensor_product, ops::BilinearTensorProductOp,
+            ops::BilinearTensorProductOpMaker, bilinear_tensor_product_grad,
+            ops::BilinearTensorProductOpGrad);
+REGISTER_OP_CPU_KERNEL(
+    bilinear_tensor_product,
+    ops::BilinearTensorProductKernel<paddle::platform::CPUPlace, float>);
```
Comment (lcy-seso) [inline paddle/operators/bilinear_tensor_product_op.cc]: register a kernel support the double type.

File: paddle/operators/bilinear_tensor_product_op.h [hunk 18]
```diff
@@ -0,0 +1,182 @@
...
+                            ctx.GetPlace());
+    auto y_scale_mat = EigenMatrix<T>::From(y_scale);
+
+    math::SetConstant<Place, T> set_zero;
+
+    // Set X@Grad be zero at first.
+    if (d_x) {
+      d_x->mutable_data<T>(ctx.GetPlace());
+      set_zero(ctx.device_context(), d_x, static_cast<T>(0));
+    }
```
Comment (lcy-seso) [inline paddle/operators/bilinear_tensor_product_op.h]:
```cpp
 if (d_x)  d_x->mutable_data<T>(ctx.GetPlace());
```

Setting zero is not necessary here.

File: paddle/operators/bilinear_tensor_product_op.h [hunk 19]
```diff
@@ -0,0 +1,182 @@
...
+
+    // Set Y@Grad be zero at first.
+    if (d_y) {
+      d_y->mutable_data<T>(ctx.GetPlace());
+      set_zero(ctx.device_context(), d_y, static_cast<T>(0));
+    }
```
Comment (lcy-seso) [inline paddle/operators/bilinear_tensor_product_op.h]:
```cpp
if (d_y)  d_y->mutable_data<T>(ctx.GetPlace());
```

File: python/paddle/v2/framework/tests/test_bilinear_tensor_product_op.py [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+            'Bias': bias,
+        }
+        self.outputs = {'Out': output + bias}
+
+    def test_check_output(self):
+        self.check_output()
+
+    def test_check_grad_normal(self):
+        self.check_grad(['X', 'Y', 'Weight', 'Bias'], 'Out')
+
```
Comment (lcy-seso) [inline python/paddle/v2/framework/tests/test_bilinear_tensor_product_op.py]: Why do we need TestBilinearTensorProductOp2 and TestBilinearTensorProductOp3?  I do not think these tests are necessary. They are not necessary boundary case for BilinearTensorProductOp.

File: paddle/operators/bilinear_tensor_product_op.cc [hunk 21]
```diff
@@ -0,0 +1,161 @@
...
+    PADDLE_ENFORCE(ctx->HasInput("X"), "Input(X) should not be null.");
+    PADDLE_ENFORCE(ctx->HasInput("Y"), "Input(Y) should not be null.");
+    PADDLE_ENFORCE(ctx->HasInput("Weight"),
+                   "Input(Weight) should not be null.");
+    PADDLE_ENFORCE(ctx->HasOutput("Out"), "Output(Out) should not be null.");
+    auto x_dims = ctx->GetInputDim("X");
+    auto y_dims = ctx->GetInputDim("Y");
+    auto weight_dims = ctx->GetInputDim("Weight");
+
+    PADDLE_ENFORCE_EQ(x_dims.size(), 2UL, "The input X must be a 2D Tensor.");
```
[the remark this answers is not in the document]
Comment (peterzhang2029) [reply paddle/operators/bilinear_tensor_product_op.cc]: Done

File: paddle/operators/bilinear_tensor_product_op.cc [hunk 22]
```diff
@@ -0,0 +1,161 @@
...
+    PADDLE_ENFORCE_EQ(y_dims.size(), 2UL, "The input Y must be a 2D Tensor.");
+    PADDLE_ENFORCE_EQ(weight_dims.size(), 3UL,
+                      "The input Weight must be a 3D tensor.");
+    PADDLE_ENFORCE(weight_dims[0],
+                   "The first dimension of Weight must be larger than 0.");
+    PADDLE_ENFORCE(weight_dims[1],
+                   "The second dimension of Weight must be larger than 0.");
+    PADDLE_ENFORCE(weight_dims[2],
+                   "The third dimension of Weight must be larger than 0.");
```
[the remark this answers is not in the document]
Comment (peterzhang2029) [reply paddle/operators/bilinear_tensor_product_op.cc]: Done

File: paddle/operators/bilinear_tensor_product_op.cc (hunk 13 above)
Comment (peterzhang2029) [reply paddle/operators/bilinear_tensor_product_op.cc]: Done

File: paddle/operators/bilinear_tensor_product_op.cc (hunk 14 above)
Comment (peterzhang2029) [reply paddle/operators/bilinear_tensor_product_op.cc]: Done

File: paddle/operators/bilinear_tensor_product_op.cc (hunk 15 above)
Comment (peterzhang2029) [reply paddle/operators/bilinear_tensor_product_op.cc]: Done

File: paddle/operators/bilinear_tensor_product_op.cc (hunk 16 above)
Comment (peterzhang2029) [reply paddle/operators/bilinear_tensor_product_op.cc]: Done

File: paddle/operators/bilinear_tensor_product_op.cc (hunk 17 above)
Comment (peterzhang2029) [reply paddle/operators/bilinear_tensor_product_op.cc]: Done

File: paddle/operators/bilinear_tensor_product_op.h (hunk 18 above)
Comment (peterzhang2029) [reply paddle/operators/bilinear_tensor_product_op.h]: There is an additive operation for d_x:
```
d_x = d_x + y_scale weight_i
```
For this reason, the elements of d_x must be initialized as 0.  Otherwise this op will lead to erroneous result.

File: paddle/operators/bilinear_tensor_product_op.h (hunk 19 above)
Comment (peterzhang2029) [reply paddle/operators/bilinear_tensor_product_op.h]: The same with d_x

File: python/paddle/v2/framework/tests/test_bilinear_tensor_product_op.py (hunk 20 above)
Comment (peterzhang2029) [reply python/paddle/v2/framework/tests/test_bilinear_tensor_product_op.py]: Removed

File: paddle/operators/bilinear_tensor_product_op.h [hunk 23]
```diff
@@ -0,0 +1,182 @@
...
+
+    // Caculate the X@Grad and Y@Grad.
+    if (d_x || d_y) {
+      Eigen::DSizes<int, 2> bcast_for_x(1, weight_dims[2]);
+      Eigen::DSizes<int, 2> bcast_for_y(1, weight_dims[1]);
+      for (int i = 0; i < weight_dims[0]; ++i) {
+        Tensor weight_i = weight->Slice(i, i + 1).Resize(
+            framework::make_ddim({weight_dims[1], weight_dims[2]}));
+        auto output_vec = d_out_mat.chip(i, 1);
+        if (d_x) {
```
Comment (lcy-seso) [inline paddle/operators/bilinear_tensor_product_op.h]: 以 dx 为例 （dy）也相同，
$dx = \frac{\partial{\mathcal{L}}}{\partial{Z}}WY^T$  其中乘以 \partial{\mathcal{L}}}{\partial{Z} 是一个broadcast 的 “scaling” 运算。

为什么不可以在 135 ~ 138 之后再进行这个 “scaling” 运算呢？这样是不是就可以直接去掉 `x_scale` 和 `x_scale` 这样两个中间变量（还有内存分配的问题）。

File: paddle/operators/bilinear_tensor_product_op.h [hunk 24]
```diff
@@ -0,0 +1,182 @@
...
+using framework::Tensor;
+
+template <typename T, int MajorType = Eigen::RowMajor,
+          typename IndexType = Eigen::DenseIndex>
+using EigenMatrix = framework::EigenMatrix<T, MajorType, IndexType>;
+
+template <typename T, int MajorType = Eigen::RowMajor,
+          typename IndexType = Eigen::DenseIndex>
+using EigenVector = framework::EigenVector<T, MajorType, IndexType>;
+
```
[the remark this answers is not in the document]
Comment (peterzhang2029) [reply paddle/operators/bilinear_tensor_product_op.h]: Done

File: paddle/operators/bilinear_tensor_product_op.h [hunk 25]
```diff
@@ -0,0 +1,182 @@
...
+    out->mutable_data<T>(ctx.GetPlace());
+
+    auto y_mat = EigenMatrix<T>::From(*y);
+    auto output_mat = EigenMatrix<T>::From(*out);
+
+    auto batch_size = x->dims()[0];
+    auto weight_dims = weight->dims();
+    auto place = ctx.GetEigenDevice<Place>();
+
+    // Create the intermediate variables.
```
[the remark this answers is not in the document]
Comment (peterzhang2029) [reply paddle/operators/bilinear_tensor_product_op.h]: Done

File: paddle/operators/bilinear_tensor_product_op.h [hunk 26]
```diff
@@ -0,0 +1,182 @@
...
+
+    auto batch_size = x->dims()[0];
+    auto weight_dims = weight->dims();
+
+    auto x_mat = EigenMatrix<T>::From(*x);
+    auto y_mat = EigenMatrix<T>::From(*y);
+    auto d_out_mat = EigenMatrix<T>::From(*d_out);
+    auto place = ctx.GetEigenDevice<Place>();
+
+    // Create the intermediate variables for gradient.
```
[the remark this answers is not in the document]
Comment (peterzhang2029) [reply paddle/operators/bilinear_tensor_product_op.h]: Done

File: paddle/operators/bilinear_tensor_product_op.h [hunk 27]
```diff
@@ -0,0 +1,182 @@
...
+                            ctx.GetPlace());
+    auto x_scale_mat = EigenMatrix<T>::From(x_scale);
+    Tensor y_scale;
+    y_scale.mutable_data<T>(framework::make_ddim({batch_size, weight_dims[2]}),
+                            ctx.GetPlace());
+    auto y_scale_mat = EigenMatrix<T>::From(y_scale);
+
+    math::SetConstant<Place, T> set_zero;
+
+    // Set X@Grad be zero at first.
```
[the remark this answers is not in the document]
Comment (peterzhang2029) [reply paddle/operators/bilinear_tensor_product_op.h]: Done

File: paddle/operators/bilinear_tensor_product_op.h [hunk 28]
```diff
@@ -0,0 +1,182 @@
...
+      set_zero(ctx.device_context(), d_x, static_cast<T>(0));
+    }
+
+    // Set Y@Grad be zero at first.
+    if (d_y) {
+      d_y->mutable_data<T>(ctx.GetPlace());
+      set_zero(ctx.device_context(), d_y, static_cast<T>(0));
+    }
+
+    // Caculate the X@Grad and Y@Grad.
```
[the remark this answers is not in the document]
Comment (peterzhang2029) [reply paddle/operators/bilinear_tensor_product_op.h]: Done

File: paddle/operators/bilinear_tensor_product_op.h (hunk 23 above)
Comment (peterzhang2029) [reply paddle/operators/bilinear_tensor_product_op.h]: 这里由于broadcast是在batch的方向展开，且TMP = scaled(X) W，scaled(X)中每一行元素所乘的放缩系数不同，所以无法在矩阵乘法之后做scaling计算。即scaled(X) W != scaled(X W).

File: paddle/operators/bilinear_tensor_product_op.h [hunk 29]
```diff
@@ -0,0 +1,182 @@
...
+              x_mat;
+          math::gemm<Place, T>(ctx.device_context(), CblasNoTrans, CblasNoTrans,
+                               batch_size, weight_dims[2], weight_dims[1], 1,
+                               x_scale.data<T>(), weight_i.data<T>(), 1,
+                               d_y->data<T>());
+        }
+      }
+    }
+
+    // Caculate the gradient of Weight.
```
[the remark this answers is not in the document]
Comment (peterzhang2029) [reply paddle/operators/bilinear_tensor_product_op.h]: Done

File: paddle/operators/bilinear_tensor_product_op.h [hunk 30]
```diff
@@ -0,0 +1,182 @@
...
+                .broadcast(bcast_for_weight) *
+            x_mat;
+        math::gemm<Place, T>(ctx.device_context(), CblasTrans, CblasNoTrans,
+                             weight_dims[1], weight_dims[2], batch_size, 1,
+                             x_scale.data<T>(), y->data<T>(), 0,
+                             d_weight_i.data<T>());
+      }
+    }
+
+    // Caculate the gradient of Bias.
```
[the remark this answers is not in the document]
Comment (peterzhang2029) [reply paddle/operators/bilinear_tensor_product_op.h]: Done

**Comment by luotao1** [comment]
There is one warning in the teamcity log, please fix it.
```
[07:08:48]W:	 [Step 1/1] In file included from /paddle/paddle/operators/bilinear_tensor_product_op.cc:15:0:
[07:08:48]W:	 [Step 1/1] /paddle/paddle/operators/bilinear_tensor_product_op.h: In instantiation of 'void paddle::operators::BilinearTensorProductKernel<Place, T>::Compute(const paddle::framework::ExecutionContext&) const [with Place = paddle::platform::CPUPlace; T = double]':
[07:08:48]W:	 [Step 1/1] /paddle/paddle/operators/bilinear_tensor_product_op.cc:159:78:   required from here
[07:08:48]W:	 [Step 1/1] /paddle/paddle/operators/bilinear_tensor_product_op.h:56:26: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
[07:08:48]W:	 [Step 1/1]      for (size_t i = 0; i < weight_dims[0]; ++i) {
[07:08:48]W:	 [Step 1/1]                           ^
[07:08:48]W:	 [Step 1/1] /paddle/paddle/operators/bilinear_tensor_product_op.h: In instantiation of 'void paddle::operators::BilinearTensorProductKernel<Place, T>::Compute(const paddle::framework::ExecutionContext&) const [with Place = paddle::platform::CPUPlace; T = float]':
[07:08:48]W:	 [Step 1/1] /paddle/paddle/operators/bilinear_tensor_product_op.cc:159:78:   required from here
[07:08:48]W:	 [Step 1/1] /paddle/paddle/operators/bilinear_tensor_product_op.h:56:26: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
```

**Comment by peterzhang2029** [comment]
Warning fixed @luotao1 @lcy-seso

File: paddle/operators/bilinear_tensor_product_op.h [hunk 31]
```diff
@@ -43,24 +43,26 @@ class BilinearTensorProductKernel : public framework::OpKernel<T> {
 
     auto batch_size = x->dims()[0];
     auto weight_dims = weight->dims();
+    int Out_dim = weight_dims[0];
```
Comment (lcy-seso) [inline paddle/operators/bilinear_tensor_product_op.h]: Out_dim --> out_dim
X_dim --> x_dim
Y_dim -->y_dim

第一个字母不要大写。

Comment (peterzhang2029) [reply paddle/operators/bilinear_tensor_product_op.h]: Done
